chia/seeder/dns_server.py

```diff
@@ -1,294 +1,365 @@
+    ipv4: List[IPv4Address]
+    ipv6: List[IPv6Address]
 
-D = None
-ns = None
-IP = "127.0.0.1"
-TTL = None
-soa_record = None
-ns_records: List[Any] = []
+    @property
+    def no_peers(self) -> bool:
+        return not self.ipv4 and not self.ipv6
 
 
-class EchoServerProtocol(asyncio.DatagramProtocol):
-    def __init__(self, callback):
-        self.data_queue = asyncio.Queue()
-        self.callback = callback
-        asyncio.ensure_future(self.respond())
+@dataclass
+class UDPDNSServerProtocol(asyncio.DatagramProtocol):
+    callback: DnsCallback
+    transport: asyncio.DatagramTransport = field(init=False)
+    data_queue: asyncio.Queue[tuple[DNSRecord, tuple[str, int]]] = asyncio.Queue()
 
-    def connection_made(self, transport):
-        self.transport = transport
+    def connection_made(self, transport: asyncio.BaseTransport) -> None:
+        # we use the #ignore because transport is a subclass of BaseTransport, but we need the real type.
+        self.transport = transport  # type: ignore[assignment]
 
-    def datagram_received(self, data, addr):
-        asyncio.ensure_future(self.handler(data, addr))
+    def datagram_received(self, data: bytes, addr: tuple[str, int]) -> None:
+        dns_request: Optional[DNSRecord] = parse_dns_request(data)
+        if dns_request is None:  # Invalid Request, we can just drop it and move on.
+            return
+        asyncio.ensure_future(self.handler(dns_request, addr))
 
-    async def respond(self):
-        while True:
+    async def respond(self) -> None:
+        while not self.transport.is_closing():
             try:
-                resp, caller = await self.data_queue.get()
-                self.transport.sendto(resp, caller)
+                reply, caller = await self.data_queue.get()
+                self.transport.sendto(reply.pack(), caller)
             except Exception as e:
-                log.error(f"Exception: {e}. Traceback: {traceback.format_exc()}.")
+                log.error(f"Exception while responding to UDP DNS request: {e}. Traceback: {traceback.format_exc()}.")
+
+    async def handler(self, data: DNSRecord, caller: tuple[str, int]) -> None:
+        r_data = await get_dns_reply(self.callback, data)  # process the request, returning a DNSRecord response.
+        await self.data_queue.put((r_data, caller))
+
+
+@dataclass
+class TCPDNSServerProtocol(asyncio.Protocol):
+    callback: DnsCallback
+    transport: asyncio.Transport = field(init=False)
+
+    def connection_made(self, transport: asyncio.BaseTransport) -> None:
+        # we use the #ignore because transport is a subclass of BaseTransport, but we need the real type.
+        self.transport = transport  # type: ignore[assignment]
 
-    async def handler(self, data, caller):
+    def data_received(self, data: bytes) -> None:
+        dns_request: Optional[DNSRecord] = parse_dns_request(data)
+        if dns_request is None:  # Invalid Request, so we disconnect and don't send anything back.
```

Comment:
We should probably log a bad request to see if we have any issues in the way clients are calling.

Reply:
Is this resolved because it is logged?